specs/ExtendedNewWindowRequested.md

```diff
@@ -0,0 +1,115 @@
+The `Name` property will return the name of the new window being created. 
+
+# Examples
+## C++: Get name of window
+
+``` cpp
+wil::com_ptr<ICoreWebView2> m_webviewEventSource;
+EventRegistrationToken m_newWindowRequestedToken = {};
+
+m_webviewEventSource->add_NewWindowRequested(
+    Callback<ICoreWebView2NewWindowRequestedEventHandler>(
+        [this](ICoreWebView2* sender, ICoreWebView2NewWindowRequestedEventArgs* args)
+          -> HRESULT 
+            {
+                wil::com_ptr<ICoreWebView2NewWindowRequestedEventArgs2> args2;
+              
+                if(SUCCEEDED(args->QueryInterface(IID_PPV_ARGS(&args2)))) 
+                {
+                    wil::unique_cotaskmem_string name;
+                    CHECK_FAILURE(args2->get_Name(&name));
+
+                    wil::unique_cotaskmem_string uri;
+                    CHECK_FAILURE(args->get_Uri(&uri));
+
+                    // Example usage of how the customer would use the window name to pass
+                    // additional context from their web content to their webview2 event handler.
+                    if (wcscmp(name.get(), L"openInNewBrowser") == 0)
+                    {
+                        ShellExecute(NULL, "open", uri.get(), 
+                            NULL, NULL, SW_SHOWNORMAL);
+                        args->put_Handled(true);
+                    }
+                    else 
+                    {
+                        HandleNewWindow(args); // App specific handling of opening a new window
+                    }
+                }
+
+              return S_OK;
+            })
+            .Get(),
+  &m_newWindowRequestedToken);
+```
+
+## C#: Get name of window
+```c#
+webView.CoreWebView2.NewWindowRequested += WebView_NewWindowRequested;
+
+void WebView_NewWindowRequested(object sender, CoreWebView2NewWindowRequestedEventArgs e)
+{
+    string newWindowName = e.Name;
+    
+    // Example usage of how the customer would use the window name to pass
+    // additional context from their web content to their webview2 event handler.
+    if (newWindowName == "openInNewBrowser")
+    {
+        Process.Start(e.Uri);
+        e.Handled = true;
+    }
+    else
+    {
+        HandleNewWindow(e); // App specific handling of opening a new window
+    }
+}
+```
+
+# API Details
+## C++
+```
+/// This is a continuation of the `ICoreWebView2NewWindowRequestedEventArgs` interface.
+[uuid(9bcea956-6e1f-43bc-bf02-0a360d73717b), object, pointer_default(unique)]
+interface ICoreWebView2NewWindowRequestedEventArgs2 : ICoreWebView2NewWindowRequestedEventArgs {
+  /// Gets the name of the new window. This window can be created via `window.open(url, windowName)`,
+  /// where the windowName parameter corresponds to `Name` property.
+  /// If no windowName is passed to `window.open`, then the `Name` property 
+  /// will be set to an empty string. Additionally, if window is opened through other means, 
```

Comment:
I don't know how to say this in the docs, but there's no empty string in WinRT (only `null`), but that does project to c# as an empty string.

Reply:
Don't need to mention it at all. By the time it reaches the projection, it has been converted to an empty string.

Interesting that Mike's point of view is that "WinRT strings can be null but cannot be empty". I look at it as a glass-half-full and say "WinRT strings cannot be null but can be empty."

I consider it half-full because `WindowsCreateString(L"")` succeeds and returns `nullptr`, instead of failing with `nullptr`. Similarly, `WindowsGetStringBuffer(nullptr)` succeeds and returns `L""` instead of failing. As far as the API is concerned, `nullptr` is a valid `HSTRING` and represents the empty string.

Reply:
Definitely not necessary to get hung up on it in this spec. But interested if there's a doc convention for explaining it.

The way I remember it is the rhyme that in WinRT there's no such thing as an empty string. I thought WindowsCreateString returned an error if you passed in ""? The doc says to pass in NULL.

I feel like there should be some standard text that says "will be empty string in C#, nullptr in C++".

Reply:
WindowsCreateString succeeds with `L""` and returns `nullptr`.

Note that this C++ API is not a WinRT API. For WebView2, the string is a cotaskmem-allocated string, never null.

Reply:
If saying either null string or empty string is reasonable for winrt let's leave it as empty string since that's what it will be for our COM & .NET APIs.